backend/ibutsu_server/scripts/mongo2postgres.py

```diff
@@ -283,89 +280,121 @@ def migrate_file(collection, Model, vprint, limit=None, filter_=None):
+    except Exception as e:
+        print(e)
+        raise e
+
+
+def migrate_widgets(mongo_url, mongo_db, postgres_url, is_verbose):
+    """Migrate all the widgets for the dashboards"""
+    try:
+        mongo, app, vprint = setup_objects(mongo_url, mongo_db, postgres_url, is_verbose)
+        with app.app_context():
+            vprint("Migrating widgets ", end="")
+            migrate_table(mongo.widgetConfig, WidgetConfig, vprint)
+    except Exception as e:
+        print(e)
+        raise e
+
+
+def migrate_projects(mongo, vprint):
+    """Migrate the projects"""
+    vprint("Migrating projects ", end="")
+    for mproject in mongo.projects.find():
+        vprint(".", end="")
+        project_id = convert_objectid_to_uuid(mproject["_id"])
+        pproject = Project.query.get(project_id)
+        if not pproject:
+            session.add(Project(id=project_id, name=mproject["name"], title=mproject["title"]))
+    session.commit()
+    session.flush()
+    vprint(" done")
 
 
-def build_mongo_connection(url):
-    """Create a MongoDB connection URL"""
-    url_parts = url.split("/")
-    database = url_parts[-1]
-    connection_url = "/".join(url_parts[:-1]) + "/?authSource={}".format(database)
-    return connection_url, database
+def migrate_tables(
+    mongo_url, mongo_db, postgres_url, is_verbose, processors, record_limit, month_limit
+):
+    """Migrate all the tables"""
+    # Set up the objects we need
+    mongo, app, vprint = setup_objects(mongo_url, mongo_db, postgres_url, is_verbose)
+    with app.app_context():
+        # first get the time range
+        sort = [("_id", -1)]
+        most_recent_record = mongo["results"].find_one(sort=sort)
+        most_recent_create_time = most_recent_record["_id"].generation_time
+        # only include most recent runs and results
+        if month_limit:
+            table_filter = {
+                "_id": {
+                    "$gt": ObjectId.from_datetime(
+                        most_recent_create_time - timedelta(days=30 * month_limit)
+                    ),
+                    "$lt": ObjectId.from_datetime(most_recent_create_time),
+                }
+            }
+        else:
+            table_filter = {"_id": {"$lt": ObjectId.from_datetime(most_recent_create_time)}}
+
+        # First, migrate the projects
+        migrate_projects(mongo, vprint)
+
+        # Now create a multiprocessing pool (it automatically uses all processors available)
+        pool = Pool(processors or None)
+        # Add the widet and reports migration functions to the pool
+        pool.apply_async(migrate_widgets, (mongo_url, mongo_db, postgres_url, is_verbose))
+        pool.apply_async(
+            migrate_reports, (mongo_url, mongo_db, postgres_url, is_verbose, table_filter)
+        )
+        # Get a list of all the runs, and add a migration job for each run to the pool
+        for run in mongo.runs.find(table_filter):
```

Comment:
In case something goes wrong during the migration, I think we should reverse sort these so that the more recent results are migrated first.